### **Ordinal Number Encoding**  
it is nothing but where we can arrange the categories according to the Particular Rank, by giving them Ranking, For e.g: Education (PHD, Master, Bachelors, High School), Grading system in a school or college, Week Days (Monday, Tuesday, Wednesday, Thursday, Friday, Saturday, Sunday)

**Importing the neccessary Libraries**

In [1]:
import datetime

Taking the todays date into a variable

In [2]:
today_date=datetime.datetime.today()

In [3]:
today_date

datetime.datetime(2020, 8, 26, 7, 7, 29, 697991)

**Datetime.timedelta is nothing but the difference between two datetime values**  
Here we are taking the difference of 3 days datetime 

In [4]:
today_date-datetime.timedelta(3)

datetime.datetime(2020, 8, 23, 7, 7, 29, 697991)

Taking the 15 days data by using `listComprehension` and `timedelta`

In [5]:
days=[today_date-datetime.timedelta(x) for x in range(0,15)]

In [6]:
days

[datetime.datetime(2020, 8, 26, 7, 7, 29, 697991),
 datetime.datetime(2020, 8, 25, 7, 7, 29, 697991),
 datetime.datetime(2020, 8, 24, 7, 7, 29, 697991),
 datetime.datetime(2020, 8, 23, 7, 7, 29, 697991),
 datetime.datetime(2020, 8, 22, 7, 7, 29, 697991),
 datetime.datetime(2020, 8, 21, 7, 7, 29, 697991),
 datetime.datetime(2020, 8, 20, 7, 7, 29, 697991),
 datetime.datetime(2020, 8, 19, 7, 7, 29, 697991),
 datetime.datetime(2020, 8, 18, 7, 7, 29, 697991),
 datetime.datetime(2020, 8, 17, 7, 7, 29, 697991),
 datetime.datetime(2020, 8, 16, 7, 7, 29, 697991),
 datetime.datetime(2020, 8, 15, 7, 7, 29, 697991),
 datetime.datetime(2020, 8, 14, 7, 7, 29, 697991),
 datetime.datetime(2020, 8, 13, 7, 7, 29, 697991),
 datetime.datetime(2020, 8, 12, 7, 7, 29, 697991)]

Taking the 15 days data into a **DataFrame**

In [7]:
import pandas as pd
data=pd.DataFrame(days)
data.columns=["Day"]

In [8]:
data.head()

Day
0 2020-08-26 07:07:29.697991
1 2020-08-25 07:07:29.697991
2 2020-08-24 07:07:29.697991
3 2020-08-23 07:07:29.697991
4 2020-08-22 07:07:29.697991

Making a new feature weekday for taking the weekdays into it

In [9]:
data['weekday'] = data['Day'].dt.day_name()
data.head()

Day    weekday
0 2020-08-26 07:07:29.697991  Wednesday
1 2020-08-25 07:07:29.697991    Tuesday
2 2020-08-24 07:07:29.697991     Monday
3 2020-08-23 07:07:29.697991     Sunday
4 2020-08-22 07:07:29.697991   Saturday

**Creating the Dictionary in which we will give ranking to the weekdays as follows**

In [10]:
dictionary={'Monday':1,'Tuesday':2,'Wednesday':3,'Thursday':4,'Friday':5,'Saturday':6,'Sunday':7
    
    
}

In [11]:
dictionary

{'Friday': 5,
 'Monday': 1,
 'Saturday': 6,
 'Sunday': 7,
 'Thursday': 4,
 'Tuesday': 2,
 'Wednesday': 3}

So the `map` function is used for mapping, means wherever ther is the weekday it will map the dictionary with it.

In [12]:
data['weekday_ordinal']=data['weekday'].map(dictionary)

**So here we can see the weekdays has been mapped, according to the rankings that we have given to them**

In [13]:
data

Day    weekday  weekday_ordinal
0  2020-08-26 07:07:29.697991  Wednesday                3
1  2020-08-25 07:07:29.697991    Tuesday                2
2  2020-08-24 07:07:29.697991     Monday                1
3  2020-08-23 07:07:29.697991     Sunday                7
4  2020-08-22 07:07:29.697991   Saturday                6
5  2020-08-21 07:07:29.697991     Friday                5
6  2020-08-20 07:07:29.697991   Thursday                4
7  2020-08-19 07:07:29.697991  Wednesday                3
8  2020-08-18 07:07:29.697991    Tuesday                2
9  2020-08-17 07:07:29.697991     Monday                1
10 2020-08-16 07:07:29.697991     Sunday                7
11 2020-08-15 07:07:29.697991   Saturday                6
12 2020-08-14 07:07:29.697991     Friday                5
13 2020-08-13 07:07:29.697991   Thursday                4
14 2020-08-12 07:07:29.697991  Wednesday                3

### **Count Or Frequency Encoding**
In this technique the Categories present into a feature will be replaced by the frequency of the Category(number) which is being used into that particular feature.

In [14]:
train_set = pd.read_csv('http://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data' , header = None,index_col=None)
train_set.head()

0                  1       2           3   ...  11  12              13      14
0  39          State-gov   77516   Bachelors  ...   0  40   United-States   <=50K
1  50   Self-emp-not-inc   83311   Bachelors  ...   0  13   United-States   <=50K
2  38            Private  215646     HS-grad  ...   0  40   United-States   <=50K
3  53            Private  234721        11th  ...   0  40   United-States   <=50K
4  28            Private  338409   Bachelors  ...   0  40            Cuba   <=50K

[5 rows x 15 columns]

**Setting the column names, as we are going to give them labels or columns names which have noT given already**  

***taking out the categorical features because we are doing feature engg for categorical variables***

In [15]:
columns=[1,3,5,6,7,8,9,13]

In [16]:
train_set=train_set[columns]

In [17]:
train_set.columns=['Employment','Degree','Status','Designation','family_job','Race','Sex','Country']

So here we can see we have taken out all the vategorical features which are present into the dataset for out further opertions

In [18]:
train_set.head()

Employment      Degree  ...      Sex         Country
0          State-gov   Bachelors  ...     Male   United-States
1   Self-emp-not-inc   Bachelors  ...     Male   United-States
2            Private     HS-grad  ...     Male   United-States
3            Private        11th  ...     Male   United-States
4            Private   Bachelors  ...   Female            Cuba

[5 rows x 8 columns]

Taking out the unique values from each and every categorical feature in our dataset,
Like how many number of unique lables present into a feature

In [19]:
for feature in train_set.columns[:]:
    print(feature,":",len(train_set[feature].unique()),'labels')

Employment : 9 labels
Degree : 16 labels
Status : 7 labels
Designation : 15 labels
family_job : 6 labels
Race : 5 labels
Sex : 2 labels
Country : 42 labels


As Country is having So many labels, So we are going to perform **Count or Frequency Encoding** on it, after that the countries will replaced by some thing else lets see.

In [20]:
country_map=train_set['Country'].value_counts().to_dict()

So here we can that in the Country the Countries has been replaced with values, as per there frequency they have been replaced with numbers like number of time the country i being repeated.

In [21]:
train_set['Country']=train_set['Country'].map(country_map) # mapping the country_map dict with Country Column
train_set.head(20)

Employment         Degree  ...      Sex Country
0           State-gov      Bachelors  ...     Male   29170
1    Self-emp-not-inc      Bachelors  ...     Male   29170
2             Private        HS-grad  ...     Male   29170
3             Private           11th  ...     Male   29170
4             Private      Bachelors  ...   Female      95
5             Private        Masters  ...   Female   29170
6             Private            9th  ...   Female      81
7    Self-emp-not-inc        HS-grad  ...     Male   29170
8             Private        Masters  ...   Female   29170
9             Private      Bachelors  ...     Male   29170
10            Private   Some-college  ...     Male   29170
11          State-gov      Bachelors  ...     Male     100
12            Private      Bachelors  ...   Female   29170
13            Private     Assoc-acdm  ...     Male   29170
14            Private      Assoc-voc  ...     Male     583
15            Private        7th-8th  ...     Male     643
16   Self-emp-not-inc        HS-grad  ...     Male   29170
17            Private        HS-grad  ...     Male   29170
18            Private           11th  ...     Male   29170
19   Self-emp-not-inc        Masters  ...   Female   29170

[20 rows x 8 columns]

### **Advantages**
1. Easy To Use
2. Not increasing feature space 

### **Disadvantages**
1. It will provide same weight if the frequencies are same for e.g (US = 100 and india = 100 same number of weight)
2. Target Guided Ordinal Encoding
3. Ordering the labels according to the target
4. Replace the labels by the joint probability of being 1 or 0

## **Target Guided Ordinal Encoding**
1. Ordering the labels according to the target
2. Replace the labels by the joint probability of being 1 or 0

In [22]:
import pandas as pd
url = "https://raw.githubusercontent.com/krishnaik06/Feature-Engineering-Live-sessions/master/titanic.csv"
df=pd.read_csv(url, usecols=['Cabin','Survived'])
df.head()

Survived Cabin
0         0   NaN
1         1   C85
2         1   NaN
3         1  C123
4         0   NaN

Replacing the NA values with **Missing** 

In [27]:
df['Cabin'].fillna('Missing',inplace=True)

In [28]:
df.head()

Survived    Cabin
0         0  Missing
1         1      C85
2         1  Missing
3         1     C123
4         0  Missing

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Survived  891 non-null    int64 
 1   Cabin     891 non-null    object
dtypes: int64(1), object(1)
memory usage: 14.0+ KB


Converting **Cabin** column into `str` data type, and then taking the first letter of each record present into **Cabin** Feature

In [40]:
df['Cabin']=df['Cabin'].astype(str).str[0]

So here we can see, that only the first letter is there into the **Cabin** Column.

In [44]:
df.head()

Survived Cabin
0         0     M
1         1     C
2         1     M
3         1     C
4         0     M

To see how many unique values does **Cabin** in having

In [45]:
df.Cabin.unique()

array(['M', 'C', 'E', 'G', 'D', 'A', 'B', 'F', 'T'], dtype=object)

In [46]:
len(df.Cabin.unique())

9

Taking the Groupby of Cabin and Survived like from **A** Cabin the chances of person survival is 46% something like that, in **each Cabin how many people got survived**

In [48]:
df.groupby(['Cabin'])['Survived'].mean()*100

Cabin
A    46.666667
B    74.468085
C    59.322034
D    75.757576
E    75.000000
F    61.538462
G    50.000000
M    29.985444
T     0.000000
Name: Survived, dtype: float64

Sorting values for taking out index

In [49]:
df.groupby(['Cabin'])['Survived'].mean().sort_values().index

Index(['T', 'M', 'A', 'G', 'C', 'F', 'B', 'E', 'D'], dtype='object', name='Cabin')

Storing into a variable

In [50]:
ordinal_labels=df.groupby(['Cabin'])['Survived'].mean().sort_values().index
ordinal_labels

Index(['T', 'M', 'A', 'G', 'C', 'F', 'B', 'E', 'D'], dtype='object', name='Cabin')

In [51]:
enumerate(ordinal_labels,0) # iterate through each and every label

Taking a `Dict Comprehension` which will give ranking to specific label as per their frequency, like for e.g **D** is having the highest frequency **75.757576** so it will get an higher rank or we can say the most higher rank

In [52]:
ordinal_labels2={k:i for i,k in enumerate(ordinal_labels,0)}
ordinal_labels2

{'A': 2, 'B': 6, 'C': 4, 'D': 8, 'E': 7, 'F': 5, 'G': 3, 'M': 1, 'T': 0}

Creating a new feature as `Cabin_ordinal_labels` and will map it with `Cabin`

In [53]:
df['Cabin_ordinal_labels']=df['Cabin'].map(ordinal_labels2)
df.head()

Survived Cabin  Cabin_ordinal_labels
0         0     M                     1
1         1     C                     4
2         1     M                     1
3         1     C                     4
4         0     M                     1

## **Mean Encoding**
Taking the mean value of each and every category present into the feature, and will be using mean value of that particular Category for further operations

here taking groupby of Cabin and Survived and then taking the mean value of it

In [58]:
mean_ordinal=df.groupby(['Cabin'])['Survived'].mean().to_dict()

In [55]:
mean_ordinal

{'A': 0.4666666666666667,
 'B': 0.7446808510638298,
 'C': 0.5932203389830508,
 'D': 0.7575757575757576,
 'E': 0.75,
 'F': 0.6153846153846154,
 'G': 0.5,
 'M': 0.29985443959243085,
 'T': 0.0}

Mapping the mean ordinal encoding with the Cabin column

In [56]:
df['mean_ordinal_encode']=df['Cabin'].map(mean_ordinal)
df.head()

Survived Cabin  Cabin_ordinal_labels  mean_ordinal_encode
0         0     M                     1             0.299854
1         1     C                     4             0.593220
2         1     M                     1             0.299854
3         1     C                     4             0.593220
4         0     M                     1             0.299854